File: scripts/case_document.py
"""The machine-readable form of a run: what `cases.build` in the MCP server persists.

The rules live in cases.py and run in the sandbox. This module only shapes their output,
so the TypeScript side never has a second implementation of triage or grouping to drift from.
"""

import hashlib

from cases import build_cases, cause_key, streaks, triage

CREDENTIAL_REASON = "catalog declares a credential is required"


def case_id(key):
    """Derived, not allocated: the same cause is the same case tomorrow."""
    return hashlib.sha1(key.encode(), usedforsecurity=False).hexdigest()[:12]


def _case(c, runs):
    return {
        "case_id": case_id(c["cause_key"]),
        "cause_key": c["cause_key"],
        "cause_kind": c["cause_kind"],
        "status_class": c["status_class"],
        "proposed_party": c["proposed_party"],
        "agency_count": len(c["members"]),
        "member_feed_ids": [d["feed_id"] for d in c["members"]],
        "corroborating": [
            {"feed_id": d["feed_id"], "reason": triage(d)} for d in c["corroborating"]
        ],
        "consecutive_runs": runs[c["cause_key"]],
    }


def case_document(detections, run_dir, run_date):
    """Everything one run says, addressed by case. Detections stay in the run file."""
    cases, singles = build_cases(detections)
    runs, prior_runs = streaks(run_dir, run_date, [c["cause_key"] for c in cases + singles])
    in_scope = [d for d in detections if d["status_class"] != "auth_declared"]
    failing = [d for d in in_scope if not d["healthy"]]
    suppressed = [
        {"feed_id": d["feed_id"], "cause_key": cause_key(d), "reason": triage(d)}
        for d in detections
        if triage(d) and not (d["healthy"] and triage(d) != CREDENTIAL_REASON)
    ]
    return {
        "run_date": run_date,
        "prior_runs_on_file": prior_runs,
        "counts": {
            "checked": len(in_scope),
            "healthy": sum(1 for d in in_scope if d["healthy"]),
            "failing": len(failing),
            "suppressed_by_credential": sum(
                1 for s in suppressed if s["reason"] == CREDENTIAL_REASON
            ),
            "suppressed_by_catalog": sum(1 for d in failing if triage(d)),
            "actionable": sum(1 for d in failing if not triage(d)),
        },
        "cases": [_case(c, runs) for c in cases],
        "individual": [_case(c, runs) for c in singles],
        "suppressed": suppressed,
    }

File: packages/mcp/src/schemas/caseDocument.ts
// The boundary between the sandbox and the server. Everything the probe emits is parsed
// here and nowhere else, so a change to the Python side fails loudly at one point rather
// than spreading `any` through the tools.
import { z } from 'zod'

import { CAUSE_KINDS, PARTY_KINDS } from '../constants/enums.js'

export const corroboratingSchema = z.object({
  feed_id: z.string(),
  reason: z.string(),
})

export const caseRecordSchema = z.object({
  case_id: z.string().regex(/^[0-9a-f]{12}$/),
  cause_key: z.string().min(1),
  cause_kind: z.enum(CAUSE_KINDS),
  status_class: z.string().min(1),
  proposed_party: z.enum(PARTY_KINDS),
  agency_count: z.number().int().nonnegative(),
  member_feed_ids: z.array(z.string()),
  corroborating: z.array(corroboratingSchema),
  consecutive_runs: z.number().int().positive(),
})

export const suppressionSchema = z.object({
  feed_id: z.string(),
  cause_key: z.string(),
  reason: z.string(),
})

export const caseDocumentSchema = z.object({
  run_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  prior_runs_on_file: z.number().int().nonnegative(),
  counts: z.object({
    checked: z.number().int().nonnegative(),
    healthy: z.number().int().nonnegative(),
    failing: z.number().int().nonnegative(),
    suppressed_by_credential: z.number().int().nonnegative(),
    suppressed_by_catalog: z.number().int().nonnegative(),
    actionable: z.number().int().nonnegative(),
  }),
  cases: z.array(caseRecordSchema),
  individual: z.array(caseRecordSchema),
  suppressed: z.array(suppressionSchema),
})

export type CaseRecord = z.infer<typeof caseRecordSchema>
export type Suppression = z.infer<typeof suppressionSchema>
export type CaseDocument = z.infer<typeof caseDocumentSchema>

/** Detection fields the server reads. The run file holds more; nothing here is a contact. */
export const detectionSchema = z.object({
  run_date: z.string(),
  observed_at: z.string(),
  feed_id: z.string().nullable(),
  provider: z.string(),
  url: z.string(),
  host: z.string(),
  path: z.string(),
  status_class: z.string(),
  healthy: z.boolean(),
  http_code: z.number().nullable(),
  content_type: z.string(),
  magic_ok: z.boolean(),
  // Optional because the runs captured before it existed are evidence and are never
  // regenerated to match a later schema.
  body_prefix: z.string().optional(),
  tls_ok: z.boolean().nullable(),
  latency_ms: z.number(),
  attempts: z.number(),
  auth_type: z.string(),
  catalog_status: z.string(),
  redirect_id: z.string(),
  contact_on_file: z.boolean(),
})

export type Detection = z.infer<typeof detectionSchema>
export const detectionsSchema = z.array(detectionSchema)
